File: TodoList-React/api/webhook.py
from http.server import BaseHTTPRequestHandler
import io
import json
import os
import random
import string
from google.oauth2 import service_account
from googleapiclient.discovery import build
from googleapiclient.http import MediaIoBaseDownload, MediaFileUpload
import requests

SCOPES = ["https://www.googleapis.com/auth/drive"]
CONFIG_FILE_ID = os.environ.get("GOOGLE_DRIVE_FILE_ID")


def get_drive_service():
    creds_json = os.environ.get("GOOGLE_SERVICE_ACCOUNT_JSON")
    creds_dict = json.loads(creds_json)
    creds = service_account.Credentials.from_service_account_info(
        creds_dict, scopes=SCOPES
    )
    return build("drive", "v3", credentials=creds)


def gerar_chave():
    letras = "".join(random.choices(string.ascii_uppercase, k=4))
    numeros = "".join(random.choices(string.digits, k=4))
    return f"{letras}{numeros}"


def enviar_whatsapp(whatsapp, matricula_nome, license_key):
    # Insira aqui a URL da sua API de WhatsApp favorita (ex: Evolution API, Z-API, etc.)
    whatsapp_api_url = os.environ.get("WHATSAPP_API_URL")
    whatsapp_token = os.environ.get("WHATSAPP_TOKEN")
    
    if not whatsapp or not whatsapp_api_url:
        return

    mensagem = (
        f"Olá, *{matricula_nome}*! Seu pagamento Pix foi confirmado com sucesso. 🚀\n\n"
        f"Sua chave de acesso ao *Gestor de Baixas* é: *{license_key}*\n\n"
        f"Guarde bem este código para realizar seus logins."
    )

    headers = {
        "Content-Type": "application/json"
    }
    if whatsapp_token:
        headers["Authorization"] = f"Bearer {whatsapp_token}"

    payload = {
        "phone": whatsapp,
        "message": mensagem
    }

    try:
        requests.post(whatsapp_api_url, json=payload, headers=headers, timeout=10)
    except Exception as e:
        print("Erro ao enviar WhatsApp:", e)


def enviar_webhook_discord(license_key, customer_email, matricula_nome, whatsapp):
    webhook_url = os.environ.get("VITE_DISCORD_WEBHOOK_URL")
    if not webhook_url:
        return

    payload = {
        "username": "Stripe Pix Bot",
        "content": "Novo pagamento Pix processado com sucesso!",
        "embeds": [{
            "title": "Nova Licença Gerada via Pix",
            "description": (
                "Um pagamento via Pix foi confirmado e a chave foi adicionada ao Drive."
            ),
            "color": int("16711680"),
            "fields": [
                {"name": "Código de Acesso", "value": license_key, "inline": True},
                {"name": "Usuário", "value": matricula_nome, "inline": False},
                {"name": "WhatsApp", "value": whatsapp if whatsapp else "Não informado", "inline": True},
                {"name": "E-mail", "value": customer_email, "inline": True},
                {"name": "Time", "value": "Agora mesmo", "inline": False},
            ],
        }],
    }

    try:
        requests.post(webhook_url, json=payload)
    except Exception as e:
        print("Erro ao enviar Webhook Discord:", e)


class handler(BaseHTTPRequestHandler):

    def do_POST(self):
        try:
            content_length = int(self.headers.get("Content-Length", 0))
            body_data = self.rfile.read(content_length)
            body = json.loads(body_data.decode("utf-8"))

            event_type = body.get("type")

            # Suporta tanto Checkout Session quanto PaymentIntent direto (Pix)
            if event_type in ["checkout.session.completed", "payment_intent.succeeded"]:
                object_data = body.get("data", {}).get("object", {})
                
                # Extrai metadados enviados na criação do Pix/Intenção
                metadata = object_data.get("metadata", {})
                matricula_nome = metadata.get("matricula_nome", "Cliente")
                whatsapp = metadata.get("whatsapp", "")
                
                # Pega o e-mail dependendo se veio de checkout ou payment_intent
                customer_email = "Cliente desconhecido"
                if "customer_details" in object_data:
                    customer_email = object_data.get("customer_details", {}).get("email", customer_email)
                elif "receipt_email" in object_data:
                    customer_email = object_data.get("receipt_email", customer_email)

                drive_service = get_drive_service()

                # 1. Baixar o Config.json atual do Google Drive
                request_file = drive_service.files().get_media(fileId=CONFIG_FILE_ID)
                fh = io.BytesIO()
                downloader = MediaIoBaseDownload(fh, request_file)
                done = False
                while not done:
                    status, done = downloader.next_chunk()

                fh.seek(0)
                config_data = json.load(fh)

                # 2. Gerar a nova chave e adicionar no array "codigos_validos"
                nova_chave = gerar_chave()
                if "codigos_validos" not in config_data:
                    config_data["codigos_validos"] = []

                config_data["codigos_validos"].append(nova_chave)

                # 3. Salvar temporariamente e atualizar no Google Drive
                temp_file_path = "/tmp/Config.json"
                with open(temp_file_path, "w", encoding="utf-8") as f:
                    json.dump(config_data, f, ensure_ascii=False, indent=2)

                media = MediaFileUpload(temp_file_path, mimetype="application/json")
                drive_service.files().update(
                    fileId=CONFIG_FILE_ID, media_body=media
                ).execute()

                # 4. Enviar notificação para o WhatsApp do usuário
                if whatsapp:
                    enviar_whatsapp(whatsapp, matricula_nome, nova_chave)

                # 5. Enviar notificação para o Discord
                enviar_webhook_discord(nova_chave, customer_email, matricula_nome, whatsapp)

                self.send_response(200)
                self.send_header("Content-type", "application/json")
                self.end_headers()
                self.wfile.write(
                    json.dumps({"status": "success", "license": nova_chave}).encode(
                        "utf-8"
                    )
                )
                return

            self.send_response(200)
            self.send_header("Content-type", "application/json")
            self.end_headers()
            self.wfile.write(
                json.dumps({"status": "ignored", "event": event_type}).encode("utf-8")
            )

        except Exception as ex:
            self.send_response(500)
            self.send_header("Content-type", "application/json")
            self.end_headers()
            self.wfile.write(
                json.dumps({"status": "error", "detalhe": str(ex)}).encode("utf-8")
            )
            return

    def do_GET(self):
        self.send_response(200)
        self.send_header("Content-type", "application/json")
        self.end_headers()
        self.wfile.write(
            json.dumps(
                {"status": "active", "message": "Webhook endpoint is running"}
            ).encode("utf-8")
        )

File: TodoList-React/api/Webhook.js
import { buffer } from 'micro';
import Stripe from 'stripe';
import { google } from 'googleapis';

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY);
const endpointSecret = process.env.STRIPE_WEBHOOK_SECRET;

export const config = {
  api: {
    bodyParser: false,
  },
};

function getDriveService() {
  const credsJson = process.env.GOOGLE_SERVICE_ACCOUNT_JSON;
  if (!credsJson) throw new Error('GOOGLE_SERVICE_ACCOUNT_JSON não definida.');

  const credsDict = typeof credsJson === 'string' ? JSON.parse(credsJson) : credsJson;

  const auth = new google.auth.GoogleAuth({
    credentials: credsDict,
    scopes: ['https://www.googleapis.com/auth/drive'],
  });

  return google.drive({ version: 'v3', auth });
}

function gerarChave() {
  const letras = Array.from({ length: 4 }, () =>
    String.fromCharCode(65 + Math.floor(Math.random() * 26))
  ).join('');
  const numeros = Array.from({ length: 4 }, () =>
    Math.floor(Math.random() * 10)
  ).join('');
  return `${letras}${numeros}`;
}

async function enviarWhatsapp(whatsapp, nome, licenseKey, dataValidadeFormatada) {
  const whatsappApiUrl = process.env.WHATSAPP_API_URL;
  const whatsappToken = process.env.WHATSAPP_TOKEN;

  if (!whatsapp || !whatsappApiUrl) return;

  const mensagem =
    `Olá, *${nome}*! Seu pagamento foi confirmado com sucesso. 🚀\n\n` +
    `Sua chave de acesso ao *Gestor de Baixas* é: *${licenseKey}*\n` +
    `Validade da licença: *${dataValidadeFormatada}* (30 dias)\n\n` +
    `Guarde bem este código para realizar seus logins.`;

  const headers = { 'Content-Type': 'application/json' };
  if (whatsappToken) {
    headers['Authorization'] = `Bearer ${whatsappToken}`;
  }

  try {
    await fetch(whatsappApiUrl, {
      method: 'POST',
      headers,
      body: JSON.stringify({ phone: whatsapp, message: mensagem }),
    });
  } catch (error) {
    console.error('Erro ao enviar WhatsApp:', error);
  }
}

async function enviarWebhookDiscord(licenseKey, customerEmail, nome, whatsapp, dataAquisicao, dataValidade) {
  const webhookUrl = process.env.VITE_DISCORD_WEBHOOK_URL;
  if (!webhookUrl) return;

  const payload = {
    username: 'Stripe Pix Bot',
    content: 'Novo pagamento processado com sucesso!',
    embeds: [
      {
        title: 'Nova Licença Gerada (30 Dias)',
        description: 'Um pagamento foi confirmado e os dados do comprador foram salvos no Drive.',
        color: 16711680,
        fields: [
          { name: 'Código de Acesso', value: licenseKey, inline: true },
          { name: 'Usuário', value: nome, inline: false },
          { name: 'WhatsApp', value: whatsapp || 'Não informado', inline: true },
          { name: 'E-mail', value: customerEmail, inline: true },
          { name: 'Data de Aquisição', value: dataAquisicao, inline: true },
          { name: 'Válido até', value: dataValidade, inline: true },
        ],
      },
    ],
  };

  try {
    await fetch(webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
  } catch (error) {
    console.error('Erro ao enviar Webhook Discord:', error);
  }
}

export default async function handler(req, res) {
  if (req.method === 'GET') {
    return res.status(200).json({ status: 'active', message: 'Webhook endpoint running' });
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Método não permitido' });
  }

  const buf = await buffer(req);
  const sig = req.headers['stripe-signature'];

  let event;

  try {
    if (endpointSecret) {
      event = stripe.webhooks.constructEvent(buf, sig, endpointSecret);
    } else {
      event = JSON.parse(buf.toString());
    }
  } catch (err) {
    console.error(`Erro no Webhook Stripe: ${err.message}`);
    return res.status(400).send(`Webhook Error: ${err.message}`);
  }

  const eventType = event.type;

  if (eventType === 'payment_intent.succeeded' || eventType === 'checkout.session.completed') {
    try {
      const objectData = event.data.object;
      const metadata = objectData.metadata || {};

      const nome = metadata.nome || metadata.matricula_nome || 'Cliente';
      const whatsapp = metadata.whatsapp || '';

      let customerEmail = 'Cliente desconhecido';
      if (objectData.customer_details?.email) {
        customerEmail = objectData.customer_details.email;
      } else if (objectData.receipt_email) {
        customerEmail = objectData.receipt_email;
      } else if (metadata.email) {
        customerEmail = metadata.email;
      }

      const fileId = process.env.GOOGLE_DRIVE_FILE_ID;
      const drive = getDriveService();

      // 1. Obter o arquivo de configuração do Drive
      const fileStream = await drive.files.get(
        { fileId, alt: 'media' },
        { responseType: 'text' }
      );

      let configData = {};
      try {
        configData = typeof fileStream.data === 'string' ? JSON.parse(fileStream.data) : fileStream.data;
      } catch (e) {
        configData = { codigos_validos: [], licencas_detalhadas: [] };
      }

      // 2. Gerar a chave e calcular as datas (30 dias)
      const novaChave = gerarChave();
      
      const agora = new Date();
      const dataValidadeObj = new Date();
      dataValidadeObj.setDate(agora.getDate() + 30);

      const dataAquisicaoISO = agora.toISOString();
      const dataValidadeISO = dataValidadeObj.toISOString();

      const dataAquisicaoFormatada = agora.toLocaleDateString('pt-BR');
      const dataValidadeFormatada = dataValidadeObj.toLocaleDateString('pt-BR');

      // 3. Atualizar array simples e salvar objeto detalhado do comprador
      if (!Array.isArray(configData.codigos_validos)) {
        configData.codigos_validos = [];
      }
      configData.codigos_validos.push(novaChave);

      if (!Array.isArray(configData.licencas_detalhadas)) {
        configData.licencas_detalhadas = [];
      }

      const registroComprador = {
        chave: novaChave,
        nome,
        email: customerEmail,
        whatsapp,
        data_aquisicao: dataAquisicaoISO,
        data_validade: dataValidadeISO,
        data_aquisicao_formatada: dataAquisicaoFormatada,
        data_validade_formatada: dataValidadeFormatada,
        dias_validade: 30,
        status: 'ativa'
      };

      configData.licencas_detalhadas.push(registroComprador);

      // 4. Salvar no Google Drive
      await drive.files.update({
        fileId,
        media: {
          mimeType: 'application/json',
          body: JSON.stringify(configData, null, 2),
        },
      });

      // 5. Enviar notificações
      await Promise.all([
        enviarWhatsapp(whatsapp, nome, novaChave, dataValidadeFormatada),
        enviarWebhookDiscord(novaChave, customerEmail, nome, whatsapp, dataAquisicaoFormatada, dataValidadeFormatada),
      ]);

      return res.status(200).json({
        status: 'success',
        license: novaChave,
        valid_until: dataValidadeFormatada
      });
    } catch (ex) {
      console.error('Erro no processamento do webhook:', ex);
      return res.status(500).json({ status: 'error', detalhe: ex.message });
    }
  }

  return res.status(200).json({ status: 'ignored', event: eventType });
    }
